### ISSUE TITLE
Compliance with Groovy 2.5's Standard Mechanism for Specifying Extension Modules

### ISSUE THREAD

**Issue by lingocoder** [gradle/gradle#11028]
### Expected Behavior

The Gradle build tool is expected to comply with the standard mechanisms that are specified for correctly configuring the external dependencies it uses. In the context of this issue „*compliance*“ and „*correctly configuring*“ means the configuration specified by the upstream external projects that produce the external libraries that the Gradle build tool depends on.

### Current Behavior

In Gradle version 5.6.2, the *`«$GRADLE_USER_HOME»/caches/5.6.2/generated-gradle-jars/gradle-api-5.6.2.jar`* file contains a service configuration file that — *according to Groovy maintainer @blackdrag* — does not comply with [*the Groovy project's most up-to-date standard mechanism for configuring Groovy Extension Modules*](http://bit.ly/Groovy8480). The same non-compliant configuration can currently be found in Gradle's version 6+ early access releases too.

### Context

As a developer of a library that will be distributed as a JPMS module, I am trying to accomplish unit testing of the classes of my JPMS-modularized library.

During the development and unit testing of my library, I was affected by the JPMS development hurdle reported in Gradle issues *#11027* and *#10408*. 

The specific Groovy project-specified standard mechanism that Gradle is not in compliance with, is prescribed [*in Groovy Issue \#8480*](http://bit.ly/Groovy8480)...

>
> ***org.codehaus.groovy.runtime.ExtensionModule should move to META-INF/groovy***
> ...
> ***Description***
> 
> „*To avoid confusing Java 9's ModuleFinder. We should fall back to the current location to handle existing jars...*“
> ...
>

I was made aware of the above issue by Groovy maintainer, @blackdrag, [*in a jigsaw-dev mailing list discussion*](http://bit.ly/jigsawdev014293)...

> ...
>„...*No idea what gradle uses here, but the standard mechanism uses now*
> *META-INF/groovy/org.codehaus.groovy.runtime.ExtensionModule*
> *(https://issues.apache.org/jira/browse/GROOVY-8480). So I assume Gradle*
> *is there still on Groovy 2.4*...“
> ...

As far as I have been able to determine, [*Gradle version 5.6+ uses Groovy version 2.5+*](http://bit.ly/buildInitL89) by default:

```
...
findLatest("groovy", "org.codehaus.groovy:groovy:(2.5,)", versionProperties)
...
```

GROOVY-8480 was resolved in March, 2018. So I would have expected Gradle to have been in compliance with Groovy's standard mechanisms for that version by now.

### The Problem

***TL;DR*** — Even if the hurdle reported in issues *#11027* and *#10408* is successfully gotten around, *`«$GRADLE_USER_HOME»/caches/5.6.2/generated-gradle-jars/gradle-api-5.6.2.jar's`* malformed *`META-INF/services/org.codehaus.groovy.runtime.ExtensionModule`* service configuration file entry means any expectation of using that particular Gradle library as an automatic module is still „*out of the question*“.

A more detailed discussion of the problem can be found in [*this jigsaw-dev mailing list thread*](http://bit.ly/jigsawdev014289).

### Steps to Reproduce 

1. [*Download and unpack this zip-archived project*](http://bit.ly/antiUpLoad)
2. In your computer's file explorer, navigate to *`«$GRADLE_USER_HOME»/caches/5.6.2/generated-gradle-jars/gradle-api-5.6.2.jar`* 
3. Open the JAR file in an archive management tool
    • Do not *extract* the contents. Actually *open* the JAR so that you can navigate it as if it were another file system folder
4. Navigate to the *`META-INF/services`* entry of the JAR
5. Extract the *`META-INF/services/java.security.Provider`* service configuration file entry to a folder on your hard drive
6. Edit the extracted *`java.security.Provider`* file so that it contains the single line shown below:
    
```
org.gradle.internal.impldep.org.bouncycastle.jce.provider.BouncyCastleProvider

```
7. Replace the original, incorrect two-line, *`META-INF/services/java.security.Provider`* service configuration file entry in the JAR file, with the corrected one-line one you just edited in Step 6.
   
8. In a terminal opened in the root folder of the project downloaded and unpacked in Step 1, run the *`:test`* task:

```
./gradlew :test
```
9. Observe the error printed to stdout:

```
...
...[QUIET] [system.out] Error occurred during initialization of boot layer
...[QUIET] [system.out] java.lang.module.FindException: Unable to derive module descriptor for «$GRADLE_USER_HOME»/caches/5.6.2/generated-gradle-jars/gradle-api-5.6.2.jar
...[QUIET] [system.out] Caused by: java.lang.module.InvalidModuleDescriptorException: Provider class moduleName=model-core not in module
...
```

### Your Environment

```
------------------------------------------------------------
Gradle 5.6.2
------------------------------------------------------------

Build time:   2019-09-05 16:13:54 UTC
Revision:     55a5e53d855db8fc7b0e494412fc624051a8e781

Kotlin:       1.3.41
Groovy:       2.5.4
Ant:          Apache Ant(TM) version 1.9.14 compiled on March 12 2019
JVM:          12 (Oracle Corporation 12+33)
OS:           Windows amd64
```

**Comment by melix** [comment]
While the fact that Gradle isn't up to date with the location of the groovy extension module file location is correct, this is completely orthogonal to JPMS. Groovy modules have _nothing_ to do with java modules.

**Comment by lingocoder** [comment]
Hi @melix,

Thanks for replying.

> „...*the groovy extension module file location...is completely orthogonal to JPMS*...“

And so it should be, melix. Just like the operation of the brakes on the handlebars of my road bike should be orthogonal to the operation of the gearing mechanism on those same handlebars. But if I *intentionally* Jerry-rig one up to use the cables and levers of the other, there is bound to be trouble at some point down the road.

> „...*Groovy modules have nothing to do with java modules*...“

I agree! That is the reason GROOVY-8480 was raised, after all. Right? 

It is true that the *goals* and *intentions* of Groovy modules have *nothing* to do with the *goals* and *intentions* of Java modules. But GROOVY-8480 says this...

> „...*To avoid confusing Java 9's ModuleFinder*...“

That issue description acknowledges that Groovy's Extension Modules' co-opting of Java's service provider-configuration file, inappropriately *Jerry-rigs the cables and levers* of the two together. While such inappropriate coupling might be a benefit to Groovy modules, it *breaks* Java modules. For that reason, 8480 is advising that using *`META-INF/services`* for Groovy modules is *the wrong thing to do*.

GROOVY-8480 is officially acknowledging, that since the two have *nothing to do with one another*, it is, therefore, inappropriate and incorrect for Groovy modules to misuse the Java modules' service provider-configuration file. 8480 advises that the correct mechanism from Groovy 2.5 onward is *`META-INF/groovy/...`*. That is: *`META-INF/services/...`* has been *`deprecated`* for Groovy 2.5+. Continuing to use it is a recipe for trouble down the road.

This is where I'm coming from as a humble user, melix...

1. You know Groovy 2.5+ has *`META-INF/groovy`* for its extension modules
2. You know Gradle 5.6+ uses Groovy 2.5+
3. You know that the use of *`META-INF/services`* is deprecated because it was *the wrong thing to do*
4. You know that there is a more up-to-date standard mechanism available
5. You know the up-to-date standard mechanism is *officially advised* to be used
6. You know that using the deprecated mechanism breaks Java modules

So that raises three simple questions:

1. Why continue to do the *wrong thing* when you know it breaks Java modules? 
2. What is the reason that *not* using Groovy's more up-to-date standard mechanism is so essentially required for 5.6+? 
3. Will Gradle *ever* update its use of Groovy extension modules to use the more up-to-date standard mechanism? 
    • If so, when? 
    • If no, why not?

From my perspective as a mere user, it seems to me that Gradle could have simply and easily upgraded to the more up-to-date standard mechanism at anytime since March 2018, and it wouldn't negatively impact Gradle's use of Groovy extension modules. Could you not? That would be an absolute win/win!

Besides, Gradle itself has its own opinions, conventions and standard mechanisms that it is always strongly advising users to comply with. It is also always deprecating things and advising users to upgrade A.S.A.P. No? So wouldn't Gradle want to, *at least*, be seen to be practicing what it preaches?

**Comment by melix** [comment]
> Why continue to do the wrong thing when you know it breaks Java modules?

Did I say we shouldn't fix? Where?

>  What is the reason that not using Groovy's more up-to-date standard mechanism is so essentially required for 5.6+?

Because this was unnoticed because nobody complained so far, as simple as that.

>    Will Gradle ever update its use of Groovy extension modules to use the more up-to-date standard mechanism?
>    • If so, when?

PRs accepted, otherwise as soon as we have bandwidth.

**Comment by lingocoder** [comment]
Hey thanks for the quick reply, @melix :+1:

>>    *Why continue to do the wrong thing when you know it breaks Java modules?*
>
> „*Did I say we shouldn't fix? Where?*“
>

You're right, melix. You never said that. Sorry if it came off like I was implying that.

>>    *What is the reason that not using Groovy's more up-to-date standard mechanism is so essentially required for 5.6+?*
>
> „*Because this was unnoticed because nobody complained so far, as simple as that.*“
>

That makes sense. Thanks, melix.

>>    *Will Gradle ever update its use of Groovy extension modules to use the more up-to-date standard mechanism?*
>>    • *If so, when?*
>
> „*PRs accepted, otherwise as soon as we have bandwidth.*“

Hey now! That's the best news I've heard all year! :1st_place_medal: 

I did consider fixing this and *#11027* myself. The other day I repeatedly searched the current remote github.com/gradle/gradle code. When I couldn't find anything there, I also grepped the hell out of my local clone of 5.6.1 I've had on my machine for a few months. But I couldn't find neither the *`META-INF/services/java.security.Policy`* file, the *`org.codehaus.groovy.runtime.ExtensionModule`* one, nor any reference or mention of either one of them anywhere in either the remote or local code base.

Can you share any pointers, melix, to where people with bandwidth should look to get started working on something like that? TIA :+1:

**Comment by lingocoder** [comment]
>> *The other day I repeatedly searched the current remote github.com/gradle/gradle code...*

These are a few of the things I bookmarked from my searching the other day...

* [*GradleImplDepsGenerationIntegrationTest.groovy*](https://github.com/gradle/gradle/blob/6fa439caa9ffd12b4ad06f7af21be237cd404f9d/subprojects/plugin-development/src/integTest/groovy/org/gradle/plugin/devel/impldeps/GradleImplDepsGenerationIntegrationTest.groovy)
* [*BaseGradleImplDepsIntegrationTest.groovy*](https://github.com/gradle/gradle/blob/6fa439caa9ffd12b4ad06f7af21be237cd404f9d/subprojects/plugin-development/src/integTest/groovy/org/gradle/plugin/devel/impldeps/BaseGradleImplDepsIntegrationTest.groovy)
* [*DefaultGeneratedGradleJarCache.java*](https://github.com/gradle/gradle/blob/b189979845c591d8c4a0032527383df0f6d679b2/subprojects/core/src/main/java/org/gradle/cache/internal/DefaultGeneratedGradleJarCache.java)
* [*DefaultCacheRepository.java*](https://github.com/gradle/gradle/blob/e975872ae2cf1d521b73f1107cae772ca09c63cd/subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/DefaultCacheRepository.java)
* [*CacheBuilder.java*](https://github.com/gradle/gradle/blob/df4b511a935359f6b84175ccc1418e472a87b03e/subprojects/persistent-cache/src/main/java/org/gradle/cache/CacheBuilder.java)

... before I gave up since none of the above references neither of those JAR file service provider-configuration entries.

**Comment by melix** [comment]
Look at https://github.com/gradle/gradle/blob/bd45f09e710845b1dca72d57e3048f334412b2e7/subprojects/model-core/src/main/resources/META-INF/services/org.codehaus.groovy.runtime.ExtensionModule

which is the extension module file reported in the original thread on jigsaw-dev.

There are, as far as I can tell, at least 5 of those files in `gradle/gradle`. It should be as easy to fix as moving them to the right location.

**Comment by rgoldberg** [comment]
https://github.com/gradle/gradle/pull/11082#issuecomment-545654889 seems to indicate that b2c3fdfdbd4872c40edf0545f2c44e2ebd41c9ae fixed this issue…

If so, should this be closed?

**Comment by ljacomet** [comment]
Closing as fixed
